File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/exam/ExamFragment.java
package com.qianlei.jiaowu.ui.fragment.exam;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.qianlei.jiaowu.MainApplication;
import com.qianlei.jiaowu.R;
import com.qianlei.jiaowu.common.Result;
import com.qianlei.jiaowu.databinding.FragmentExamBinding;
import com.qianlei.jiaowu.entity.Examination;
import com.qianlei.jiaowu.utils.DateUtil;

import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * 考试的fragment
 *
 * @author qianlei
 */
public class ExamFragment extends Fragment implements AdapterView.OnItemSelectedListener, SwipeRefreshLayout.OnRefreshListener {

    private ExamViewModel examViewModel;
    private FragmentExamBinding binding;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_exam, container, false);

        ViewModelProvider.AndroidViewModelFactory factory = new ViewModelProvider.AndroidViewModelFactory(MainApplication.getInstance());
        examViewModel = factory.create(ExamViewModel.class);
        binding.termChooseView.setItemSelectedListener(this);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        binding.swipeRefreshLayout.setOnRefreshListener(this);
        examViewModel.getExamListData().observe(this.getViewLifecycleOwner(), this::updateExam);
        binding.setLifecycleOwner(this);
        return binding.getRoot();
    }

    /**
     * 更新考试信息
     *
     * @param result 考试信息内容
     */
    private void updateExam(@NotNull Result<List<Examination>> result) {
        if (result.isSuccess()) {
            ExamAdapter adapter = new ExamAdapter(MainApplication.getInstance(), result.getData());
            binding.recyclerView.setAdapter(adapter);
        } else {
            Toast.makeText(getContext(), result.getMsg(), Toast.LENGTH_SHORT).show();
        }
        binding.swipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        examViewModel.changeTerm(binding.termChooseView.getYear(), binding.termChooseView.getTerm());
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
        examViewModel.changeTerm(String.valueOf(DateUtil.getCurYear()), String.valueOf(DateUtil.getCurTerm()));
    }

    @Override
    public void onRefresh() {
        examViewModel.refreshData(binding.termChooseView.getYear(), binding.termChooseView.getTerm());
    }
}

File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/exam/ExamFragment.kt
package com.qianlei.jiaowu.ui.fragment.exam

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.qianlei.jiaowu.R
import com.qianlei.jiaowu.common.Result
import com.qianlei.jiaowu.entity.Examination
import com.qianlei.jiaowu.entity.Term
import kotlinx.android.synthetic.main.fragment_exam.*

/**
 * 考试的fragment
 *
 * @author qianlei
 */
class ExamFragment : Fragment(), OnItemSelectedListener, OnRefreshListener {
    private val examViewModel by viewModels<ExamViewModel>()

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_exam, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        termChooseView.setItemSelectedListener(this)
        recyclerView.layoutManager = LinearLayoutManager(context)
        swipeRefreshLayout.setOnRefreshListener(this)
        swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary)
        examViewModel.examData.observe(this.viewLifecycleOwner, Observer { result: Result<List<Examination>> -> updateExam(result) })
    }


    /**
     * 更新考试信息
     *
     * @param result 考试信息内容
     */
    private fun updateExam(result: Result<List<Examination>>) {
        val c = context ?: return
        if (result.isSuccess()) {
            val data = result.data
            val adapter = ExamAdapter(c, data)
            recyclerView.adapter = adapter
        } else {
            Toast.makeText(context, result.msg, Toast.LENGTH_SHORT).show()
        }
        swipeRefreshLayout.isRefreshing = false
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        swipeRefreshLayout.isRefreshing = true
        examViewModel.changeTerm(termChooseView.term)
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {
        swipeRefreshLayout.isRefreshing = true
        examViewModel.changeTerm(Term.getNowTerm())
    }

    override fun onRefresh() {
        examViewModel.refreshData(termChooseView.term)
    }
}